qiskit/circuit/library/n_local/qaoa_ansatz.py

```diff
@@ -0,0 +1,235 @@
+            `arXiv:1411.4028 <https://arxiv.org/pdf/1411.4028>`_
+    """
+    def __init__(self,
+                 cost_operator: 'OperatorBase',
+                 reps: int,
+                 initial_state: Optional[QuantumCircuit] = None,
+                 mixer_operator: Optional[Union[QuantumCircuit, 'OperatorBase']] = None,
+                 name: str = "qaoa"):
+        r"""
+        Args:
+            cost_operator: The operator representing the cost of the optimization problem,
+                denoted as :math:`U(C, \gamma)` in the original paper.
+            reps: The integer parameter p, which determines the depth of the circuit,
+                as specified in the original paper.
+            initial_state: An optional initial state to use. If `None` is passed then a set of
+                Hadamard gates is applied as an initial state to all qubits.
+            mixer_operator: An optional custom mixer to use instead of the global X-rotations,
+                denoted as :math:`U(B, \beta)` in the original paper. Can be an operator or
+                an optionally parameterized quantum circuit.
+            name: A name of the circuit, default 'qaoa'
+        Raises:
+            TypeError: invalid input
+        """
+        super().__init__(name=name)
+        self._cost_operator = cost_operator
+        self._reps = reps
+        self._initial_state = initial_state
+        self._mixer_operator = mixer_operator
+
+        # set this circuit as a not-built circuit
+        self._num_qubits = None
+        self._num_parameters = 0
+        self._bounds = None
+        self._mixer = None
+        self._data = None
+
+    def _check_configuration(self, raise_on_failure: bool = True) -> bool:
+        valid = True
+
+        if self._cost_operator is None:
+            valid = False
+            if raise_on_failure:
+                raise AttributeError("The operator representing the cost of "
+                                     "the optimization problem is not set")
+        if self._reps is None or self._reps < 1:
+            valid = False
+            if raise_on_failure:
+                raise AttributeError("The integer parameter p, which determines the depth "
+                                     "of the circuit, either not set or set to non-positive value")
+
+        num_qubits = self._cost_operator.num_qubits
+        if self._initial_state is not None and self._initial_state.num_qubits != num_qubits:
+            valid = False
+            if raise_on_failure:
+                raise AttributeError("The number of qubits of the initial state does not match "
+                                     "the number of qubits of the cost operator")
+
+        if self._mixer_operator is not None and self._mixer_operator.num_qubits != num_qubits:
+            valid = False
+            if raise_on_failure:
+                raise AttributeError("The number of qubits of the mixer does not match "
```

Comment:
It might be nice to have the number of qubits in each case in the message.

Reply:
Agree.